x/ccv/provider/keeper/partial_set_security.go

```diff
@@ -67,3 +74,171 @@ func (k Keeper) HandleOptOut(ctx sdk.Context, chainID string, providerAddr types
+		if !validator.IsBonded() {
+			continue
+		}
+
+		m[pubKey.String()] = abci.ValidatorUpdate{
+			PubKey: pubKey,
+			Power:  k.stakingKeeper.GetLastValidatorPower(ctx, valAddress),
+		}
+	}
+
+	for _, addr := range validatorAddressesToRemove {
+		_, pubKey, err := k.getValAddressAndPublicKey(ctx, addr)
+		if err != nil {
+			continue
+		}
+
+		m[pubKey.String()] = abci.ValidatorUpdate{
+			PubKey: pubKey,
+			Power:  0,
+		}
+	}
+
+	var out []abci.ValidatorUpdate
+	for _, update := range m {
+		out = append(out, update)
+	}
+
+	// similarly to `AccumulateChanges`, we sort validators for determinism
+	sort.Slice(out, func(i, j int) bool {
+		if out[i].Power != out[j].Power {
+			return out[i].Power > out[j].Power
+		}
+		return out[i].PubKey.String() > out[j].PubKey.String()
+	})
+
+	return out
+}
+
+// ComputeNextValidators computes the next validator set that is responsible for validating on a consumer chain.
+// The returned opted-in validators correspond to the next `currentValidators`.
+func (k Keeper) ComputeNextValidators(ctx sdk.Context,
+	currentValidators []OptedInValidator,
+	validatorAddressesToAdd []types.ProviderConsAddress,
+	validatorAddressesToRemove []types.ProviderConsAddress,
+) []OptedInValidator {
+	isRemoved := make(map[string]bool)
+	for _, val := range validatorAddressesToRemove {
+		isRemoved[val.ToSdkConsAddr().String()] = true
+	}
+
+	var out []OptedInValidator
+	for _, val := range currentValidators {
+		if isRemoved[val.ProviderAddr.ToSdkConsAddr().String()] {
+			continue
+		}
+		valAddress, _, err := k.getValAddressAndPublicKey(ctx, val.ProviderAddr)
+		if err != nil {
+			continue
+		}
+
+		val.Power = uint64(k.stakingKeeper.GetLastValidatorPower(ctx, valAddress))
+		if val.Power == 0 {
+			continue
+		}
+		out = append(out, val)
+	}
+
+	for _, addr := range validatorAddressesToAdd {
+		valAddress, _, err := k.getValAddressAndPublicKey(ctx, addr)
+		if err != nil {
+			continue
+		}
+
+		validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, addr.ToSdkConsAddr())
+		if !found {
+			continue
+		}
+		if !validator.IsBonded() {
+			continue
+		}
+
+		power := uint64(k.stakingKeeper.GetLastValidatorPower(ctx, valAddress))
+		// validator just opted in and hence sets `BlockHeight` as the current height
+		out = append(out, OptedInValidator{ProviderAddr: addr, BlockHeight: uint64(ctx.BlockHeight()), Power: power})
+	}
+
+	return out
+}
+
+// ResetCurrentValidators resets the opted-in validators with the newest set that was computed by
+// `ComputePartialSetValidatorUpdates` and hence this method should only be called after
+// `ComputePartialSetValidatorUpdates` has completed. Method also clears all the `ToBeOptedIn` and `ToBeOptedOut` states.
+func (k Keeper) ResetCurrentValidators(ctx sdk.Context, chainID string, nextValidators []OptedInValidator) {
+	k.DeleteAllOptedIn(ctx, chainID)
+	for _, val := range nextValidators {
+		k.SetOptedIn(ctx, chainID, val.ProviderAddr, val.BlockHeight, val.Power)
+	}
```

Comment:
Seems very inneficient to delete all opten in entries and set the new ones. If in an epoch there will be no changes in the voting powers, this will be a very expensive way to leave the state unchanged.

Reply:
Is my understanding correct that at this stage, nothing is written to disk? Only if the transactions succeed, at the end all the changes would get written. If that's the case, why's this expensive? Wouldn't this correspond to a simple write in memory?
I agree that most likely the validator set won't change between epochs but probably it's unlikely that powers would remain the same between epochs for all validators, although this depends on how long an epoch is. Even in this case, I would need to loop through all currently opted in validators and compare with the `nextValidators` to see if a change took place.

Reply:
might be premature optimization to worry about this? especially once we have epochs this happens rarely and it's at most ~200 entries being deleted and written per consumer